### PR TITLE
[feature] GitHub as datasource

### LINKED ISSUE #1788: [experimental] add github datasource
**Summary**

Add github as one datasource, for github repo data analysis, like star history and eveything you want.

```
create database [org_name] engine=github;
use [org_name];
select * from [repo_name];
```

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Add `github_table` as storage engine, can get statistical data from github API.
usage:
```sql
MySQL> create database github;  -- no matter what database you use
MySQL> use github;
MySQL>create table datafuselabs engine=github; -- create a table, name is a owner's name you concern
MySQL> select * from datafuselabs;
+------------------------+----------+------------+------------+-------------+----------------+-------------------+-------------------+----------------------+
| reposiroty             | language | license    | star_count | forks_count | watchers_count | open_issues_count | subscribers_count | created_at           |
+------------------------+----------+------------+------------+-------------+----------------+-------------------+-------------------+----------------------+
| .github                |          | apache-2.0 |          0 |           0 |              0 |                 1 |                 4 | 2021-11-13T14:20:07Z |
| databend               | Rust     | apache-2.0 |       2530 |         227 |           2530 |               308 |                64 | 2020-10-10T06:04:42Z |
| databend-playground    | Shell    | apache-2.0 |          2 |           0 |              2 |                 2 |                 6 | 2021-10-19T06:43:03Z |
| datafuse-operator      | Go       | apache-2.0 |          7 |           0 |              7 |                11 |                 2 | 2021-04-01T23:45:56Z |
| datafuse-presentations | HTML     | No License |         26 |          12 |             26 |                 0 |                 7 | 2021-06-03T10:17:23Z |
| datafuse-shop          |          | apache-2.0 |          9 |           1 |              9 |                 0 |                 4 | 2021-06-29T17:03:38Z |
| fusebots               | Go       | apache-2.0 |          9 |           3 |              9 |                 7 |                 3 | 2021-05-20T05:49:11Z |
| test-infra             | Go       | apache-2.0 |          6 |           6 |              6 |                 7 |                 3 | 2021-06-02T04:01:42Z |
| weekly                 |          | apache-2.0 |         16 |           5 |             16 |                 0 |                 5 | 2021-07-23T13:20:32Z |
+------------------------+----------+------------+------------+-------------+----------------+-------------------+-------------------+----------------------+
9 rows in set (15.23 sec)
Read 9 rows, 899 B in 15.226 sec., 0.5911031885553668 rows/sec., 59.04 B/sec.
```
image

## Changelog

- New Feature

## Related Issues

Fixes #1788 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: query/src/datasources/table/github/data_accessor.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use common_exception::Result;
+use reqwest::header::USER_AGENT;
+use reqwest::Client;
+
+use super::items::RepoDetail;
+use super::items::RepoMetaSimple;
+
+const USER_AGENT_VAL: &str = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.71 Safari/537.36";
+
+pub async fn get_owner_repositories_details(owner: &str) -> Result<Vec<RepoDetail>> {
```
Comment (BohuTANG) [inline query/src/datasources/table/github/data_accessor.rs:24]: FYI, there is a rust github client `octocrab`:
https://github.com/XAMPPRocky/octocrab

Comment (Veeupup) [reply query/src/datasources/table/github/data_accessor.rs:24]: it seems so nice to use this crate here, many thanks

**Comment by BohuTANG** [comment]
Cool :+1: 
There is an improvement here:
1. database name is the orgs name
2. repo name is the table name

For example:
```
create databse datafuselabs; -- the database is org name
use datafuselabs;
show tables; -- show all the repos info
select * from databend; -- get the databend repo info from the github
```

A more in-depth optimization is, we can get a repo pr table like:
```
use datafuselabs;
select * from databend_pr; -- databend is the repo, pr is the pr category
```

A comments table:
```
use datafuselabs;
select * from databend_comment; -- databend is the repo, comment is the comment category
```

**Comment by BohuTANG** [comment]
I believe this is a good example of how to add your datasource

**Comment by Veeupup** [comment]
> Cool 👍 There is an improvement here:
> 
> 1. database name is the orgs name
> 2. repo name is the table name
> 
> For example:
> 
> ```
> create databse datafuselabs; -- the database is org name
> use datafuselabs;
> show tables; -- show all the repos info
> select * from databend; -- get the databend repo info from the github
> ```
> 
> A more in-depth optimization is, we can get a repo pr table like:
> 
> ```
> use datafuselabs;
> select * from databend_pr; -- databend is the repo, pr is the pr category of the repo
> ```
> 
> A comments table:
> 
> ```
> use datafuselabs;
> select * from databend_comment; -- databend is the repo, comment is the comment category of the repo
> ```

Your advice really make sense! appreciate your example. But for now maybe not that easy to support the above example usage.

The example above has a problem that we can not figure it out whether the database `datafuselabs` is for github analysis or normal usage because standard create database SQL syntax is not supported to specify a engine.

So we can only create a table with `github engine` for now(not for database), we can not create a database whose name is organization name to analyze organization statistics.

If we want to finish the above SQL example, I think we can finish it from the below way:
Support a unique create database SQL syntax, maybe like `create database datafuselabs usage=github`. And then we can clearly know that this database `datafuselabs` will be used for github analysis and in turn we can create many tables by the organization's repos.

If anything I miss, please inform me of that.

**Comment by BohuTANG** [comment]
Oh, i forgot one important thing: we have removed the database engine in https://github.com/datafuselabs/databend/pull/2291
and we would not want them back again :(

File: query/src/datasources/table/github/github_table.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        _ctx: DatabendQueryContextRef,
+        plan: &ReadDataSourcePlan,
+    ) -> Result<SendableDataBlockStream> {
+        let limit = plan
+            .push_downs
+            .as_ref()
+            .and_then(|extras| extras.limit)
+            .unwrap_or(100) as u8;
+
+        let arrays = self.get_repos_array(limit).await?;
```
Comment (sundy-li) [inline query/src/datasources/table/github/github_table.rs:150]: This limit did not respect sort plan.

Comment (Veeupup) [reply query/src/datasources/table/github/github_table.rs:150]: thank you for reminding of that, I will fix it later.

**Comment by BohuTANG** [comment]
/lgtm

Let's merge first, and consider add database engine or not.

**Comment by BohuTANG** [comment]
API changed, reverted from the main.

**Comment by Veeupup** [comment]
@BohuTANG  I wil fix confilcts first and request a new pr to merge it into master before #2925 ?

**Comment by BohuTANG** [comment]
> @BohuTANG I will fix confilcts first and request a new pr to merge it into master before #2925 ?

I will finish https://github.com/datafuselabs/databend/issues/2925 next week, after it done, we can add a database engine='github' and this PR is more interesting then, what do you think?

**Comment by Veeupup** [comment]
> > @BohuTANG I will fix confilcts first and request a new pr to merge it into master before #2925 ?
> 
> I will finish #2925 next week, after it done, we can add a database engine='github' and this PR is more interesting then, what do you think?

work as you expect.
many thanks for your consideration, as it is an experimental feature, I'm not hurry to finish it.

**Pull request merged**